### PR TITLE
Row-based read support

### LINKED ISSUE #41: Add read support + high level API for reading Parquet files
Current read API is low-level, and you can deserialize column values and repetition and definition levels separately. We should add more of high level API that allows to map column batches into rows (structs or vectors).

### LINKED ISSUE #42: Add CLI tools for reading files
Add CLI tools like `parquet-mr/parquet-tools` as binaries, so people could use them to inspect files. We have already started doing it for file schema/metadata, we should add something to list records in a file, e.g. in JSON or tabular format or something similar.

### PR DESCRIPTION
**Pull request by sadikovi**
This PR adds new API for row-based reading of Parquet files.

This is one of the layers of API that will be finally provided (with a prospect of dropping record API in favour of Arrow, once it is integrated): 
- Low-level column readers API (_already exists_)
- Record API (_this PR_)
- Arrow API (_TODO_)

Most of the code is in `src/record` module:
- `api.rs` contains all necessary API for mapping to `Row`, which is our internal representation of records, supports most of the primitive types (easy to add new ones) and all complex types, e.g. structs, lists, maps.
- `triplet.rs` contains internal triplet iterator that provides access to `(value, def level, rep level)` tuple, performs necessary buffering and spacing of values.
- `reader.rs` contains the code to assemble a tree of readers and to traverse the tree. Also contains `RecordIter` and `RowIter` iterators of records.

API for `FileReader` and `RowGroupReader` is also updated to return an iterator of `Row`s:
- `fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<RowIter>;`
- `fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<RowIter>;`

These methods (especially for `FileReader`) can take optional projected schema to read only certain columns from a file. When `None` is provided, full schema is implied.
`RowIter` will automatically load the next row group, as iterator progresses. For `RowGroupReader` row iterator will only load data from that row group.

I also added `parquet-read.rs` file in `bin` directory, to read Parquet files. This allows to quickly inspect files similar to `parquet-mr/parquet-tools`. Now we have two binaries - one is for inspecting schema `parquet-schema` and another one for reading data `parquet-read`.

There are also some minor updates, like making certain methods public so they can be used in reader and moving common test methods into `test_common.rs`.

Added the following test files:
- `nonnullable.impala.parquet` (Impala test file)
- `nullable.impala.parquet` (Impala test file)
- `nulls.snappy.parquet` (file that contains nulls only)
- `nested_lists.snappy.parquet` (file with nested lists)
- `nested_maps.snappy.parquet` (file with nested maps)

> I think it is a good idea to update main README with available APIs - will do it after this PR is merged.

Closes #41 
Closes #42

### CODE REVIEW

**Comment by sadikovi** [comment]
@sunchao could you review the final version of record reader? It has not changed much, but I made some comment updates and minor tweaks. 

Thanks!

File: src/bin/read-file.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+use std::fs::File;
+use std::path::Path;
+use std::process;
+
+use parquet::file::reader::{FileReader, SerializedFileReader};
+
+fn main() {
+  let args: Vec<String> = env::args().collect();
+  if args.len() != 2 && args.len() != 3 {
+    println!("Usage: read-file <file-path> <num-records>");
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read-file.rs]: Yes, will change.

File: src/bin/read-file.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+extern crate parquet;
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/bin/read-file.rs]: I can rename them to `parquet-read` and `parquet-schema`. Will that work?

File: src/file/reader.rs [hunk 3]
```diff
@@ -50,6 +51,11 @@ pub trait FileReader {
   /// the same as this. Otherwise, the row group metadata stored in the row group reader
   /// may outlive the file reader.
   fn get_row_group(&self, i: usize) -> Result<Box<RowGroupReader>>;
+
+  /// Get full iterator of `Row` from a file (over all row groups).
+  /// Projected schema can be a subset of or equal to the file schema, when it is None,
+  /// full file schema is assumed.
+  fn get_row_iter(&self, projection: Option<SchemaType>) -> Result<FileRowIter>;
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/file/reader.rs]: Yes, we should be able to, will change.

File: src/record/api.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+  /// Converts BYTE_ARRAY into either UTF8 string or array of bytes.
+  pub fn convert_byte_array(
+    physical_type: PhysicalType, logical_type: LogicalType, value: ByteArray
+  ) -> Self {
+    match physical_type {
+      PhysicalType::BYTE_ARRAY => {
+        match logical_type {
+          LogicalType::UTF8 | LogicalType::ENUM | LogicalType::JSON => {
+            Row::Str(String::from_utf8(value.data().to_vec()).unwrap())
```
[the remark this answers is not in the document]
Comment (sadikovi) [reply src/record/api.rs]: It is marked as unsafe. I can add it like this: 
```rust
let value = unsafe { String::from_utf8_unchecked(value.data().to_vec()) };
Row::Str(value)
```

Is it okay?

File: src/bin/read-file.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply src/bin/read-file.rs]: Yes, that should be better 👍

File: src/record/api.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply src/record/api.rs]: Yes I think it should be OK.

**Comment by sadikovi** [comment]
@sunchao I addressed your comments. Could you do another pass on this PR? Thanks!

**Comment by sadikovi** [comment]
Thanks for the following up on the build error!

As you mentioned on JIRA, it looks like we need to fix `try_from` crate. I suggest we do that, otherwise it would fail for other people that use the latest nightly Rust. Let me know if you are planning to open a PR in the `try_from` repository, or if you want I could try fixing it.

**Comment by sunchao** [comment]
NP @sadikovi . Yes I'm working on a PR for the `try_from` repo - will post soon :) 
Meanwhile you can change the `travis.yml` to use `nightly-2018-03-14` to unblock the merge.

**Comment by sunchao** [comment]
Actually the fix is more involved than I expected. I think it might be wiser to wait until there's a conclusion in the [Rust thread](https://github.com/rust-lang/rust/pull/49305). It is breaking multiple libraries.

**Comment by sunchao** [comment]
It failed again (not sure why since it succeeded on my machine) - can you try `nightly-2018-03-26`?

**Comment by sadikovi** [comment]
@sunchao I updated, thanks!

**Comment by sunchao** [comment]
Merged! Thanks @sadikovi for this awesome contribution!

**Comment by sadikovi** [comment]
@sunchao Thanks for merging! I really appreciate your review and help throughout working on read support. 

I will open another pull request with README update, including how to read schema and/or files.

**Pull request merged**